File: meeting-booking-api/src/main/java/com/alejandro/meetingbooking/service/ReservationService.java
package com.alejandro.meetingbooking.service;

import com.alejandro.meetingbooking.dto.request.ReservationRequest;
import com.alejandro.meetingbooking.dto.response.AvailabilityResponse;
import com.alejandro.meetingbooking.dto.response.ReservationResponse;

import java.time.LocalDate;
import java.util.List;

public interface ReservationService {
    ReservationResponse createReservation(ReservationRequest request);
    void cancelReservation(Long reservationId);
    ReservationResponse findReservation(Long reservationId);
    List<ReservationResponse> findAll();
    AvailabilityResponse getAvailability(Long roomId, LocalDate date);
}

File: meeting-booking-frontend/src/app/services/reservation-service.ts
import {Injectable, inject} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Reservation} from '../models/reservation.model';
import {ReservationRequest} from '../models/reservation-request.model';
import {AvailabilityResponse} from '../models/availability-response.model';

@Injectable({
  providedIn: 'root'
})

export class ReservationService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = 'http://localhost:8080/api/v1/reservations';

  getAllReservations(){
    return this.http.get<Reservation[]>(this.apiUrl);
  }

  getReservationById(id: number){
    return this.http.get<Reservation[]>(`${this.apiUrl}/${id}`);
  }

  createReservation(reservation: ReservationRequest){
    return this.http.post<Reservation>(this.apiUrl, reservation);
  }

  cancelReservation(id: number){
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getAvailability(roomId: number, date: string) {
    return this.http.get<AvailabilityResponse>(
      `${this.apiUrl}/${roomId}/availability`,
      {
        params: {
          date
        }
      }
    );
  }
}

